lib/qmp.ml

```diff
@@ -99,126 +99,156 @@ module Event = struct
-      | x -> failwith (Printf.sprintf "unknown command %s" x)
-    ))
-  | `Assoc list when List.mem_assoc "return" list ->
-    let id = if List.mem_assoc "id" list then Some (string (List.assoc "id" list)) else None in
-    (match List.assoc "return" list with
-      | `Assoc [] -> Success (id, Unit)
-      | `Assoc list when List.mem_assoc "status" list ->
-        Success (id, Status (string (List.assoc "status" list)))
-      | `Assoc list when List.mem_assoc "enabled" list
-                      && List.mem_assoc "auth" list && List.mem_assoc "family"  list
-                      && List.mem_assoc "service" list && List.mem_assoc "host" list ->
-        Success (id, Vnc (
-          let enabled = bool (List.assoc "enabled" list)  in
-          let auth = string (List.assoc "auth" list) in
-          let family = string (List.assoc "family" list) in
-          let service = int_of_string (string (List.assoc "service" list)) in
-          let host = string (List.assoc "host" list) in
-          {enabled; auth; family; service; host}))
-      | `Assoc list when List.mem_assoc "product-num" list
-                      && List.mem_assoc "build-num" list -> (
-        try
-          Success (id, Xen_platform_pv_driver_info (
-            let product_num = int (List.assoc "product-num" list) in
-            let build_num = int (List.assoc "build-num" list) in
-            {product_num; build_num}))
-        with e ->
-          Error(None, { cls = "JSONParsing"; descr = (Printf.sprintf "%s:%s" (Printexc.to_string e) x) })
-        )
-      | `Assoc list when List.mem_assoc "enabled" list ->
-        let enabled = bool (List.assoc "enabled" list) in
-        let present = bool (List.assoc "present" list) in
-        Success (id, Enabled {enabled; present})
-      | `List ((`Assoc pair :: _) as list) when List.mem_assoc "name" pair ->
-        Success (id, Name_list (List.map (function
-                             | `Assoc [ "name", `String x ] -> x
-                             | _ -> failwith "assoc") list))
-      | `Assoc list when List.mem_assoc "fdset-id" list -> (
-        try
-          Success (id, Fd_info (
-            let fd = int (List.assoc "fd" list) in
-            let fdset_id = int (List.assoc "fdset-id" list) in
-            {fd; fdset_id}))
-        with e ->
-          Error(None, { cls = "JSONParsing"; descr = (Printf.sprintf "%s:%s" (Printexc.to_string e) x) })
+  in
+
+  let event json =
+    let ts x = json |> U.member "timestamp" |> U.member x |> U.to_int in
+    { timestamp = (ts "seconds", ts "microseconds")
+    ; event     = json |> U.member "event" |> U.to_string
+    }
+  in
+
+  let execute json =
+    let arguments = U.member "arguments" in
+    let eject json =
+      let device = json |> arguments |> U.member "device" |> U.to_string in
+      let force  = json |> arguments |> U.member "force"  |> function `Null -> None | force -> Some (force |> U.to_bool) in
```

Comment:
Take a look at `U.to_bool_option` and likewise at `U.to_string_option` and so on below. It already does what you need.